Really disable blammo if no configuration is specified

PaulMougel opened the pull request:
This Pull Request modifies Logger's behavior:
- it makes Logger.level() return Levels.OFF if no level is set. Previously, this returned undefined.
- it disables Logger.traceEntry() and Logger.traceExit() if Logger.level() < Levels.TRACE.

Together, these two commits make sure that no log will take place if blammo isn't configured, thus not impacting the application's performance.

Here is my specific use case : I have an application that uses SaXPath, which itself uses blammo. I don't want any blammo logging operation to occur, as it really slows down my application.

Commit:
Logger.getLevel() returns OFF if this.level isn't set.

This is useful especially if blammo isn't configured, ie. if the project doesn't contain a configuration file. In this case, we assume that no log should be done.

Commit:
Logger.traceEntry() and Logger.traceExit() don't do anything if Logger.level < TRACE

This commit makes both functions' behavior consistent with Logger.log().

StevenLooman commented:
Hi Paul,

Thank you for your pull request. Returning undefined for Logger.level() should have been no issue. In Javascript, undefined > 0 (or undefined > 60, for that matter) evaluates to false. Nevertheless, it improves readability, so it's a good thing to change.

Disabling traceEntry/traceExit might not be what we want though. Consider this case: at the normal/running state of a program the logger has level OFF. Patching a method usually happens during the init state, so the logger will never patch/hook into the call. Then, in the program I flip some switch to turn on debugging. Now, the call will never be logged, because the method is never patched. What are you trying to achieve here?

For the problems you are fixing, can you make a test case which shows that is was broken? Also, the current test cases are broken (see https://travis-ci.org/StevenLooman/blammo/builds/10535010). Can you fix this?

Thanks!

Steven

Commit:
Fix 40b19a557573a5f6faa80395b6e4b8ea2569b4f3

PaulMougel commented:
Sorry about the broken tests, I backported the changes really quickly and made a dumb mistake, it [should be fixed](https://travis-ci.org/StevenLooman/blammo/jobs/10613997).

I see your point about turning logging on/off at runtime.

In my very XML intensive application I noticed using [node-webkit-agent](https://github.com/c4milo/node-webkit-agent) that more than 30% of my CPU's time was spent in blammo's logging, which is a bit annoying.

Here is the _Heavy (Bottom Up)_ view from the inspector:
![perf_heavv](https://f.cloud.github.com/assets/1788596/1024537/5d95732e-0e28-11e3-9a22-486b73c451c0.png)
notice the calls to Date, which almost only come from blammo.

And some stats extracted from the _Tree_ view:
`SaXPath.onCloseTag` : 17.46%, and inside this function call `Logger.traceEntry` : 17.14% (which means 0.32% of the total CPU time is taken by the actual XML parsing!)
`SaXPath.onOpenTag`: 14.01% ; `Logger.traceEntry`: 13.74%

With the two (now three) commits above, XML processing time was almost divided by two.

As a side note, `undefined` doesn't eval as > 0 nor 60, at least in the v8 version used by node 0.8

```
> undefined > 0
false
> undefined > 60
false
```

Commit:
Logging level checks for Logger.traceEntry() and Logger.traceExit() are now done at runtime

This allows to dynamically enable or disable the methode tracing by calling Logger.setLevel().

StevenLooman commented:
Thanks for fixing. Thankfully, Travis CI caught the failing test. :)

You are right about the `undefined > 0` evaluating. I misread the Logger.log method, I'll merge that fix.

17%/13% is way too much! I didn't know it would eat that much time! I'll release a new version of SaXPath which only does the `blammo.Logger.traceEntry` call in debug/develop mode. In fact, I'll make Blammo a develop dependency.

Can you undo commit 40b19a5 and update the pull request? Then I'll merge it immediately, and later today (after 18h, probably) fix SaXPath.

PaulMougel commented:
I just pushed a commit that makes the logging level check at runtime, what do you think about it? I'll need to benchmark it more but it seems not to slow down the application that much.
On the topic on making Blammo a develop dependency: I indeed think it could be nice, as we don't usually need to trace every method call in every library used in a program..

StevenLooman commented:
The pull request still touches traceEntry/traceExit. Can you remove/undo the commit/change/

StevenLooman commented:
In the mean time, I've released SaXPath 0.5.4 which has no longer the blammo dependency (at all.)

Commit:
Revert 19fb1e3d44363a8d25c7ebab8faa051d1ae26767, f8d592897234b0d2a735eca04b20ccbe44f694e2, 40b19a557573a5f6faa80395b6e4b8ea2569b4f3

PaulMougel commented:
Alright, this should undo the changes made to traceEntry and traceExit.

StevenLooman commented:
Thanks! Merged.

StevenLooman commented:
Oops, wrong button :)

StevenLooman commented:
Just released 0.4.5
